### PR TITLE
feat(modal): PureRender support custom footer callback

### PR DESCRIPTION
**Pull request by Wxh16144**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     Modal PureRender support custom footer callback      |
| 🇨🇳 Chinese |    --       |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 475e635</samp>

Added a new `PurePanel` component that can render custom modal panels with custom footers. Refactored `render-panel.tsx` and `footer` prop handling to support this feature.

### <samp>🤖 Generated by Copilot at 475e635</samp>

*  Rename `components/modal/demo/render-panel.tsx` to `components/modal/PurePanel.tsx` and export `PurePanel` component as a new feature ([link](https://github.com/ant-design/ant-design/pull/44985/files?diff=unified&w=0#diff-93d352b81d4be73b9052974d0ac2171fd5d5c9ec35ad5765511cc21078e3b081L2-R18))
*  Modify `PurePanelProps` interface to accept a custom footer render function as a prop ([link](https://github.com/ant-design/ant-design/pull/44985/files?diff=unified&w=0#diff-9103d37c6668043c170813d54cb43ca28bc6cf111eedfc6718aeae7f540a0b27L16-R17))
*  Destructure and omit `footer` prop from `panelProps` object and use it by `Footer` component instead ([link](https://github.com/ant-design/ant-design/pull/44985/files?diff=unified&w=0#diff-9103d37c6668043c170813d54cb43ca28bc6cf111eedfc6718aeae7f540a0b27R31), [link](https://github.com/ant-design/ant-design/pull/44985/files?diff=unified&w=0#diff-9103d37c6668043c170813d54cb43ca28bc6cf111eedfc6718aeae7f540a0b27L63-R64))
*  Add an example of using `PurePanel` with a custom footer render function to the demo file ([link](https://github.com/ant-design/ant-design/pull/44985/files?diff=unified&w=0#diff-93d352b81d4be73b9052974d0ac2171fd5d5c9ec35ad5765511cc21078e3b081R30-R41))

### CODE REVIEW

File: components/modal/PurePanel.tsx [hunk 1]
```diff
@@ -28,6 +28,7 @@ const PurePanel: React.FC<PurePanelProps> = (props) => {
     type,
     title,
     children,
+    footer /** omit */, // eslint-disable-line @typescript-eslint/no-unused-vars
```
Comment (yoyo837) [inline components/modal/PurePanel.tsx:31]: 个人建议用 `eslint-disable-next-line`并放到上方, 因为右边的注释很容易因为格式化工具改变位置.

Comment (yoyo837) [reply components/modal/PurePanel.tsx:31]: 不过, 解构丢弃的字段好像可以配置 `no-unused-vars`的lint选项, 让他不报错, 不知道是否有必要, 也与当前PR无关.

Comment (Wxh16144) [reply components/modal/PurePanel.tsx:31]: 我本来放在上面的， 发现右边也可以 哈哈哈哈哈，所以就直接提交了

Comment (crazyair) [reply components/modal/PurePanel.tsx:31]: 按理说，下面有 ...rest，这里应该不会有报错才对

Comment (crazyair) [reply components/modal/PurePanel.tsx:31]: eslint 这条规则，要加下 ignoreRestSiblings: true
```
'@typescript-eslint/no-unused-vars': [2, { args: 'none', ignoreRestSiblings: true }]
```

File: components/modal/PurePanel.tsx [hunk 2]
```diff
@@ -60,7 +61,7 @@ const PurePanel: React.FC<PurePanelProps> = (props) => {
     additionalProps = {
       closable: closable ?? true,
       title,
-      footer: props.footer === undefined ? <Footer {...props} /> : props.footer,
+      footer: <Footer {...props} />,
```
Comment (afc163) [inline components/modal/PurePanel.tsx:64]: 这样会导致 `footer={null}` 无法隐藏页脚。

Comment (Wxh16144) [reply components/modal/PurePanel.tsx:64]: 是的，那我保持和 modal 的逻辑一致

File: components/modal/PurePanel.tsx [hunk 3]
```diff
@@ -28,6 +28,8 @@ const PurePanel: React.FC<PurePanelProps> = (props) => {
     type,
     title,
     children,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (crazyair) [inline components/modal/PurePanel.tsx:31]: 这个可以去掉的吧，应该不会报错

Comment (Wxh16144) [reply components/modal/PurePanel.tsx:31]: 去不了呀，会报 ts 错误

Comment (Wxh16144) [reply components/modal/PurePanel.tsx:31]: thx

**Review by MadCcc — APPROVED** [review]

File: components/modal/PurePanel.tsx [hunk 4]
```diff
@@ -60,7 +61,7 @@ const PurePanel: React.FC<PurePanelProps> = (props) => {
     additionalProps = {
       closable: closable ?? true,
       title,
-      footer: props.footer === undefined ? <Footer {...props} /> : props.footer,
+      footer: footer !== null && <Footer {...props} />,
```
Comment (kiner-tang) [inline components/modal/PurePanel.tsx:64]: 如果是 `false` 的话是不是也不应该渲染 footer ?

Comment (Wxh16144) [reply components/modal/PurePanel.tsx:64]: copy modal 的逻辑，ref： https://github.com/ant-design/ant-design/blob/b3cb6b30ed38c10d46f591186cdf53e5bb6fcd26/components/modal/Modal.tsx#L103-L105

Comment (kiner-tang) [reply components/modal/PurePanel.tsx:64]: 感觉是不是得改一下， false 也是 合法的 ReactNode，而用户传 false 的预期肯定是不想展示的，就像 React 里面的 {false} 也直接不渲染一样

Comment (Wxh16144) [reply components/modal/PurePanel.tsx:64]: footer 有用例支撑的  ref：https://github.com/ant-design/ant-design/blob/b3cb6b30ed38c10d46f591186cdf53e5bb6fcd26/components/modal/__tests__/Modal.test.tsx#L44-L52

> 而用户传 false 的预期肯定是不想展示的, rc-dialog 也有 case 的

https://github.com/react-component/dialog/blob/bb69ed382561d0fa8a65975ef1854389bae6a0ef/src/Dialog/Content/Panel.tsx#L81-L84

Comment (Wxh16144) [reply components/modal/PurePanel.tsx:64]: 不过 PureRender 没有测试用例，毕竟这个组件是不推荐开发者使用的，所以如果使用出现问题，antd 不管 😁

**Pull request merged**